### DISCUSSION TITLE
Confused about HTML template rendering from `pb.js`

### DISCUSSION THREAD

**Discussion by benstigsen** [pocketbase/pocketbase#5761] [Q&A] [answered]
I'm familiar with Go templates, but when rendering them from JavaScript the behaviour is confusing.
I have this HTML:
```html
<html>
  <head></head>
  <body>
    {{ range .posts }}
      <p>{{ . }}</p>
    {{ end }}
  </body>
</html>
```

And I have this JavaScript:
```js
routerAdd("GET", "/posts", (c) => {
  const renderer = $template.loadFiles(
    `${__hooks}/views/home.html`
  );

  const posts = arrayOf(
    new DynamicModel({
      title: "",
      published: "",
    }),
  );

  $app.dao().db()
    .newQuery("SELECT title, published FROM posts WHERE public = TRUE")
    .all(posts);

  return c.html(200, renderer.render({ posts: posts }));
});
```

This results in the following HTML body:
```html
<p>{Foo }</p>
```

There is no published date, so that makes sense, but why is it not like the object?
I can't do the following:
```html
{{ range .posts }}
  <p>{{ .title }}</p>
{{ end }}
```
Because the `title` field doesn't exist?
If I `console.log()` it in my JavaScript:
```js
  $app.dao().db()
    .newQuery("SELECT title, published FROM posts WHERE public = TRUE")
    .all(posts);

  console.log(JSON.stringify(posts));

  return c.html(200, renderer.render({ posts: posts }));
```

Then I get:
```
[{"title":"Foo","published":""}]
```

I've tried also doing:
```js
return c.html(200, renderer.render(posts))
{{ index . 0 }} // { Foo}
```

And:
```js
return c.html(200, renderer.render({ ...posts })
```

None of them seem to work like I expect them to. I must be doing something wrong, but this is what I'm trying to do:
```html
{{ range . }}
  <p>{{ .title }}</p>
  <p>{{ .published }}</p>
{{ end }}
```

**Comment by ganigeorgiev** [comment, marked as answer]
I haven't run the above example but from a quick look `<p>{{ .title }}</p>` most likely wouldn't work because `title` is not a property of a plain map/object (POJO). It is a field of `DynamicModel`, which is essentially a Go struct and public Go struct's fields are always capitalized. Or in other words, replacing `{{ .title }}` with `{{ .Title }}` should probably work.

An alternative option (but not very efficient) could be to try to serialize the `posts` into a plain array of objects, aka. "JSON.parse(JSON.stringify(posts))" and `.title` in this case should work because it will be a map key.

_As a side-note, I'm planning writting a custom and more simpler (with minimal DSL) Go template engine in the near future and will consider the JSVM case too._

**Reply by benstigsen** [reply]
Ah right! I always forget that, even when using Go regularly, the camel vs PascalCase access difference.

**Comment by benstigsen** [comment]
Alright, now could I ask for something slightly unrelated?
I have a "posts" collection where each post contains:
- title
- content
- published
- tags `select multiple`

For using tags, in the "Working with Relations" documentation, each post has a `relations multiple` to a tag. Internally is there any difference? Is `select multiple` just stored as a JSON blob?

**Reply by ganigeorgiev** [reply]
Each multivalued field type (multiple `file`, `select`, `relation`) stores its value as serialized JSON array so there is no difference regarding that but there is a difference in the filter behavior and the supported operations between the `select` and `relation` type fields.

If you plan of having a lot tags (eg. 20+) I recommend using a dedicated "tags" collection and a `relation` field pointing from/to it. It will be more efficient, will have more control over who can list/view the tags, can add extra attributes (e.g. description), etc.

The `select` field was intended mostly for enums-like fields, e.g. `status`, fixed `roles`, etc.

**Reply by benstigsen** [reply]
Thank you!
I'm really enjoying PocketBase, you've made the design very clean, it's also intuitive to use and very easy to extend. I imagine I'll be using it a lot!
